File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        
        level_order = []

        storage = deque([root])

        while storage:
            level_length = len(storage)
            level = []
            for i in range(level_length):
                node = storage.popleft()
                if node:
                    level.append(node.val)
                    storage.append(node.left)
                    storage.append(node.right)
            
            if level:
                level_order.append(level)
        
        return level_order

File: Data Structures & Algorithms/binary-tree-diameter/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        // NOTE: current node is NOT part of the diameter itself
        int diameter{0};

        dfs(root, diameter);

        return diameter;
    }

private: 
    int dfs(TreeNode* root, int& diameter){
        if (!root) return 0;

        int left = dfs(root->left, diameter);
        int right = dfs(root->right, diameter);

        // calculating diameter is max of left + right
        diameter = std::max(diameter, left + right);
        
        // when you return up to parent, do 1 + max(left + right)
        return 1 + std::max(left, right);

    }

};
